librubyfmt/src/ruby.rs

```diff
@@ -81,29 +83,82 @@ extern "C" {
     pub fn rb_id2name(id: ID) -> *const libc::c_char;
     pub fn rb_ary_entry(arr: VALUE, idx: libc::c_long) -> VALUE;
     pub fn rb_raise(cls: VALUE, msg: *const libc::c_char);
+    pub fn rb_block_call(
+        obj: VALUE,
+        method_id: ID,
+        argc: libc::c_int,
+        argv: *const VALUE,
+        block: extern "C" fn(_: VALUE, _: VALUE, _: libc::c_int, _: *const VALUE) -> VALUE,
+        outer_scope: VALUE,
+    ) -> VALUE;
 }
 
 pub fn current_exception_as_rust_string() -> String {
+    let ruby_string = unsafe { eval_str("$!.inspect") }.expect("Error evaluating `$!.inspect`");
+    unsafe { ruby_string_to_str(ruby_string) }.to_owned()
+}
+
+macro_rules! intern {
+    ($s:literal) => {
+        rb_intern(concat!($s, "\0").as_ptr() as _)
+    };
+}
+
+// Safety: This function expects an initialized Ruby VM capable of evaling code
+pub unsafe fn eval_str(s: &str) -> Result<VALUE, ()> {
+    let rubyfmt_program_as_c = CString::new(s).expect("unexpected nul byte in Ruby code");
+    let mut state = 0;
+    let v = rb_eval_string_protect(rubyfmt_program_as_c.as_ptr(), &mut state);
+    if state != 0 {
+        Err(())
+    } else {
+        Ok(v)
+    }
+}
+
+extern "C" fn real_debug_inspect(v: VALUE) -> VALUE {
     unsafe {
-        let res = eval_str("$!.inspect").expect("this can't fail");
-        let ptr = rubyfmt_rstring_ptr(res);
-        let length = rubyfmt_rstring_len(res);
-        String::from_raw_parts(ptr as _, length as _, length as _)
+        let inspect = rb_funcall(v, intern!("inspect"), 0, std::ptr::null() as *const VALUE);
+        let char_pointer = rb_string_value_cstr(&inspect) as *mut i8;
+        let cstr = CStr::from_ptr(char_pointer);
+        let s = cstr.to_str().expect("it's utf8");
+        debug!("{}", s);
+        Qnil
     }
 }
 
-pub fn eval_str(s: &str) -> Result<VALUE, ()> {
+pub fn debug_inspect(v: VALUE) {
     unsafe {
-        let rubyfmt_program_as_c = CString::new(s).expect("it should become a c string");
         let mut state = 0;
-        let v = rb_eval_string_protect(
-            rubyfmt_program_as_c.as_ptr(),
-            &mut state as *mut libc::c_int,
-        );
+        rb_protect(real_debug_inspect as _, v, &mut state);
         if state != 0 {
-            Err(())
-        } else {
-            Ok(v)
+            let s = current_exception_as_rust_string();
+            panic!("blew us: {}", s);
         }
     }
 }
+
+pub fn raise(s: &str) {
+    let cstr = CString::new(s).expect("it's not null");
```

Comment:
"it's not null" doesn't accurately describe this error condition. Can we just handle a nul byte being contained in the string gracefully or at least give a useful error message?